velox/connectors/hive/PartitionIdGenerator.cpp

```diff
@@ -21,44 +21,135 @@ namespace facebook::velox::connector::hive {
+    } else {
+      uint64_t nextPartitionId = partitionIds_.size();
+      VELOX_USER_CHECK_LT(
+          nextPartitionId,
+          maxPartitions_,
+          "Exceeded limit of distinct partitions.");
+
+      partitionIds_.emplace(valueId, nextPartitionId);
+      savePartitionKeys(nextPartitionId, input, i);
+
+      result[i] = nextPartitionId;
+    }
+  }
+
+  recentMaxId_ = *std::max_element(result.begin(), result.end());
+}
+
+void PartitionIdGenerator::computeValueIds(
+    const RowVectorPtr& input,
+    raw_vector<uint64_t>& valueIds) {
   allRows_.resize(input->size());
   allRows_.setAll();
 
-  auto partitionVector = input->childAt(hasher_->channel())->loadedVector();
-  hasher_->decode(*partitionVector, allRows_);
+  bool rehash = false;
+  for (auto& hasher : hashers_) {
+    auto partitionVector = input->childAt(hasher->channel())->loadedVector();
+    hasher->decode(*partitionVector, allRows_);
+    if (!hasher->computeValueIds(allRows_, valueIds)) {
+      rehash = true;
+    }
+  }
+
+  if (rehash) {
+    uint64_t multiplier = 1;
+    for (auto& hasher : hashers_) {
+      multiplier = hasher->enableValueIds(multiplier, 50);
+
+      VELOX_CHECK_NE(
+          multiplier,
+          exec::VectorHasher::kRangeTooLarge,
+          "Number of requested IDs is out of range.");
+    }
 
-  if (!hasher_->computeValueIds(allRows_, result)) {
-    uint64_t range = hasher_->enableValueIds(1, kHasherReservePct);
-    VELOX_CHECK_NE(
-        range,
-        exec::VectorHasher::kRangeTooLarge,
-        "Number of requested IDs is out of range.");
+    for (auto& hasher : hashers_) {
+      bool ok = hasher->computeValueIds(allRows_, valueIds);
+      VELOX_CHECK(ok);
+    }
 
-    VELOX_CHECK(
-        hasher_->computeValueIds(allRows_, result),
-        "Cannot assign new value IDs.");
+    updateValueToPartitionIdMapping();
   }
+}
 
-  recentMaxId_ = *std::max_element(result.begin(), result.end());
-  maxId_ = std::max(maxId_, recentMaxId_);
+void PartitionIdGenerator::updateValueToPartitionIdMapping() {
+  if (partitionIds_.empty()) {
+    return;
+  }
+
+  auto numPartitions = partitionIds_.size();
 
-  VELOX_USER_CHECK_LE(
-      maxId_, maxPartitions_, "Exceeded limit of distinct partitions.");
+  partitionIds_.clear();
+
+  raw_vector<uint64_t> newValueIds(numPartitions);
+  SelectivityVector rows(numPartitions);
+  for (auto i = 0; i < hashers_.size(); ++i) {
+    auto& hasher = hashers_[i];
+    hasher->decode(*partitionKeys_->childAt(hasher->channel()), rows);
+    bool ok = hasher->computeValueIds(rows, newValueIds);
+    VELOX_CHECK(ok);
+  }
+
+  for (auto i = 0; i < numPartitions; ++i) {
+    partitionIds_.emplace(newValueIds[i], i);
+  }
+}
+
+void PartitionIdGenerator::savePartitionKeys(
+    uint64_t partitionId,
+    const RowVectorPtr& input,
+    vector_size_t row) {
+  for (auto i = 0; i < partitionChannels_.size(); ++i) {
+    auto channel = partitionChannels_[i];
+    partitionKeys_->childAt(i)->copy(
```

Comment:
This line `PartitionIdGenerator::savePartitionKeys()`,

and the following line in `PartitionIdGenerator::updateValueToPartitionIdMapping()`
```c++
for (auto i = 0; i < hashers_.size(); ++i) {
  hasher->decode(*partitionKeys_->childAt(hasher->channel()), rows);
}
```

and the following line in contructor
```c++
  for (auto channel : partitionChannels_) {
    partitionKeyTypes.push_back(inputType->childAt(channel));
  }
```
should be consistent in whether the child vectors of `partitionKeys_` are indexed by partition ordinal or partition channel.

Given the VectorHasher is initialized with partition channel as below, I guess we have to use the partition channel as index?
```c++
  for (auto channel : partitionChannels_) {
    hashers_.emplace_back(
        exec::VectorHasher::create(inputType->childAt(channel), channel));
  }
```

Reply:
Nice catch. I fixed the bug and added a test case to expose it. We should use ordinal when accessing partitionValues_:

```
hasher->decode(*partitionValues_->childAt(i), rows);
```